### PR TITLE
separate init and docking

### PR DESCRIPTION
**Pull request by Feriolet**
As discussed, the initialisation and docking process will be separated to make it easier to implement future features. The steps for separating it and enable reusing database are shown below:

1. If a user submits an input file without config.yml only init step is occurred. Setup table is not populated.

> I only added setup['yaml'] so restore_setup_db function can still take in previous args in the initialisation step

2. If a user submits config.yml the setup table is populated and docking is started.

> The setup table content in docking will overwrite the content in initialisation step, by DELETE and INSERT command.
3. If a user submits both an input file and config.yml both steps occur sequentially.

4. Revise the section related to restoring settings from DB if it already exists. It is used for continuation of interrupted docking.

> Edited the DB to conditionally trigger setup['config'] related python lines.
5. Command line arguments should be separated into two (argparse) groups related to these two major steps.

> done with add_argument_group

6. Implement a script “make_clean_copy.py” which takes a DB and removes all data obtained during the docking step to get a DB copy with prepared structures only to be used for docking with another program. After removal of data use VACUUM query to reduce the DB size.

### CODE REVIEW

File: easydock/run_dock.py [hunk 1]
```diff
@@ -131,17 +131,23 @@ def main():
                                                  'will be ignored with the exception of hostfile, dask_report, ncpu '
                                                  'and verbose.',
                                      formatter_class=RawTextArgumentDefaultsHelpFormatter)
-    parser.add_argument('-i', '--input', metavar='FILENAME', required=False, type=filepath_type,
+    input_output_group = parser.add_argument_group('Input/output files')
+    init_group = parser.add_argument_group('Initialization parameters')
+    docking_group = parser.add_argument_group('docking parameter groups')
```
Comment (DrrDom) [inline easydock/run_dock.py:136]: Change the name to `Docking parameters`

File: easydock/run_dock.py [hunk 2]
```diff
@@ -131,17 +131,23 @@ def main():
...
+    dock_n_init_group = parser.add_argument_group('Applicable for both docking and initialization parameters')
```
Comment (DrrDom) [inline easydock/run_dock.py:137]: Change the name to `Common arguments`

**Comment by DrrDom** [comment]
It took me a while to remember the original logic of setup restore and understand the changes. The logic seems not perfect.

1. Delete setup and rewrite it completely looks dangerous. Can we lose some info if a user will not supply some previous args or supply new args values? I tried to run different experiments, it looked working, but my gut feeling is alerted :)
I would suggest to update only NULL fields with `populate_setup_db`. Would it be better?
2. It seems that we do not need to add `'protein', 'protein_setup'` to line 146 database.py
`if colname in ['config', 'protein', 'protein_setup']:`
Why do we need this?
3. For file copying in make_clean_copy.py please use Python module.
```python
import shutil
shutil.copyfile(src, dst)
```
4. I expect to clean the setup table, keeping the `yaml` field
5. Grouping of arguments:
- input/ouput - OK
- init
   - max_stereoisomers
   - protonation
   - no_tautomerization
   - prefix
- docking
   - program
   - config
   - sdf
   - hostfile
   - dask_report
   - tmpdir
- common
   - ncpu
   - verbose

**Comment by Feriolet** [comment]
1. Yes, removing and rewriting the setup column also does not feel right to me as well. I can't remember why I decided not to use UPDATE instead. I think it is because UPDATE in sql requires each column to be written individually, and I am not sure how to accomodate it for custom argument in `args_to_save`. The `config` column also needs to be updated because it only stores initialisation parameters during initialisation step, where the docking config is turned to default null value.
2. the code loops the `protein` and `protein_setup`  stored in `values` variable in line 145        `for colname, value in values.items():`. 

Removing the `protein` and `protein_setup` will give `KeyError: 'During loading no relevant argument name was found to match the loading field name: protein'`. 

4. I am not sure I understand this point haha. 

I will fix the other issues first.

**Comment by DrrDom** [comment]
1. Config column should not be updated, because it contains only docking setup. Probably you meant yaml column. However, changes in yanl column may cause some changes, if we continue interrupted calculations the content of yaml field will be re-written and the path to the original config.yml file will be lost (a temporary path will be created during continuation and it will replace the original path in yaml field). This will not cause an error, but is something inaccurate.
2. Yes, it would be better to check for None `if colname in ['config'] or value is None:`
3. OK
4. I mean in make_clean_copy to not remove yaml field, because we will lost information about settings at the preparation step.
5. Seems OK

There are possible issues with allowing to change config and program. If a user will run docking, interrupt it, change config/program and continue calculations, the second part of dockings should be altered and disagree with the first part. I did not test this, but I expect such behavior from the code and previously implemented logic which allowed to replace only args affecting only calculation running/distribution.
Config (and protein and protein setup) can be added to DB only once and afterwards it should not be possible to change these. We can check whether config field is empty in DB and program is NULL, if so, keep corresponding input args, otherwise discard them. However, it still looks complicated to be easy to reuse in other programs, e.g. crem-dock.  Currently I cannot propose a better solution.

**Comment by Feriolet** [comment]
Oh, yes, I meant the yaml column my bad. In case of config, I agree that it should only be updated once. This also means that I should also delete the setup config column in the `make_clean_copy` file. 

For yaml content, maybe we can similarly only update it once the config parameter is not null anymore in the yaml column?

**Comment by DrrDom** [comment]
> Oh, yes, I meant the yaml column my bad. In case of config, I agree that it should only be updated once. This also means that I should also delete the setup config column in the `make_clean_copy` file.

Yes, only ymal column will survive
 
> For yaml content, maybe we can similarly only update it once the config parameter is not null anymore in the yaml column?

This seems correct.

File: easydock/make_clean_copy.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                                    dock_time = NULL,
+                                    time = NULL   """)
+    conn.commit()
+
+    res = cur.execute('SELECT * FROM setup')
+    removed_colnames = [d[0] for d in res.description if d[0] != 'yaml']
+    remove_col_sql_line = 'UPDATE setup SET '
+    for colnames in removed_colnames:
+        remove_col_sql_line += colnames + ' = NULL, '
+    remove_col_sql_line = remove_col_sql_line[:-2]
```
Comment (DrrDom) [inline easydock/make_clean_copy.py:25]: This looks more expressive and clearer

```python
sql = 'UPDATE setup SET ' +  ','.join(f'{item} = NULL' for item in remove_colnames)
```

**Comment by DrrDom** [comment]
Regarding `populate_setup_db`. I would suggest to check in the beginning if the config field is not NULL and run the subsequent code otherwise exit the function. It should be more straightforward and easier to understand. Because I looked at the function and only at the end I understood that all this code will have sense if the config is NULL.

File: easydock/make_clean_copy.py [hunk 4]
```diff
@@ -18,13 +18,11 @@ def create_clean_db_copy(db_fname, new_db_fname):
...
 
     res = cur.execute('SELECT * FROM setup')
-    removed_colnames = [d[0] for d in res.description if d[0] != 'yaml']
-    remove_col_sql_line = 'UPDATE setup SET '
-    for colnames in removed_colnames:
-        remove_col_sql_line += colnames + ' = NULL, '
-    remove_col_sql_line = remove_col_sql_line[:-2]
+    remove_colnames = [d[0] for d in res.description if d[0] != 'yaml']
+    sql = 'UPDATE setup SET ' +  ','.join(f'{item} = NULL' for item in remove_colnames)
+    sql = sql[:-2]
```
Comment (DrrDom) [inline easydock/make_clean_copy.py:23]: Is this line necessary? It seems not, because there will be no trailing symbols. Or am I mistaken?

Comment (Feriolet) [reply easydock/make_clean_copy.py]: yes, you're right. There is no comma at the end of the string line anymore, so it should be removed

**Comment by DrrDom** [comment]
Thank you! I'll test changes locally and merge PR if everything will be fine. I created the `dev` branch, where we will merge these and further changes before the release.

**Comment by DrrDom** [comment]
All tests were passed. Thank you!

**Pull request merged**